agents/craftassist/craftassist_swarm_master.py

```diff
@@ -73,28 +87,150 @@ def init_controller(self):
+        query = "SELECT MEMORY FROM Task WHERE ((running>=1) AND (paused <= 0))"
+        _, task_mems = self.memory.basic_search(query)
+        if not task_mems:
+            time.sleep(sleep_time)
+            return
+        for mem in task_mems:
+            if "swarm_worker_0" not in mem.get_tags():
+                continue
+            mem.task.step()
+            if mem.task.finished:
+                mem.update_task()
+
     def assign_task_to_worker(self, i, task_name, task_data):
+        cur_task = TASK_MAP[task_name](self, task_data)
+        self.memory.tag(cur_task.memid, "swarm_worker_{}".format(i))
         if i == 0:
-            TASK_MAP[task_name](self, task_data)
+            pass
+        else:
+            self.swarm_workers[i-1].input_tasks.put((task_name, task_data, cur_task.memid)) 
+
+    def handle_memory_query(self, query):
+        query_id = query[0]
+        query_name = query[1]
+        query_args = query[2:]
+        if query_name in self.handle_query_dict.keys():
+            to_return = self.handle_query_dict[query_name](*query_args)
         else:
-            self.swarm_workers[i-1].input_tasks.put((task_name, task_data))
+            logging.info("swarm master cannot handle memory query: {}".format(query))
+            raise NotImplementedError
+        to_return = self.safe_object(to_return)
+        return tuple([query_id, to_return])
 
+    def safe_single_object(self, input_object):
+        if is_picklable(input_object):
+            return input_object       
+        all_attrs = dir(input_object)
+        return_obj = empty_object()
+        for attr in all_attrs:
+            if attr.startswith("__"):
+                continue
+            if type(getattr(input_object, attr)).__name__ in dir(__builtins__):
+                setattr(return_obj, attr, getattr(input_object, attr))
+        return return_obj
+
+    def safe_object(self, input_object):
+        if isinstance(input_object, tuple):
+            tuple_len = len(input_object)
+            to_return = []
+            for i in range(tuple_len):
+                to_return.append(self.safe_single_object(input_object[i]))
+            return tuple(to_return)
+        else:
+            return self.safe_single_object(input_object)
+        
     def start(self):
         # count forever unless the shutdown signal is given
         for swarm_worker in self.swarm_workers:
             swarm_worker.start()
 
+        init_status = [False] * (self.num_agents - 1)
+
         while not self._shutdown:
             try:
-                self.step()
+                if all(init_status):
+                    self.step()
+                for i in range(self.num_agents-1):
+                    flag = True
+                    # TODO: implement perception memory --> handle the perceiptions queue
+                    
+                    # task updates info from swarm worker process
```

Comment:
could we collect this into a method (just for readability), like "update_tasks_with_worker_data" ?